### PR TITLE
Add [As]VarULE impls for Vec<T>/[T]

### PR DESCRIPTION
**Pull request by Manishearth**
On request by @zbraniecki

This lets us nest vectors (one level deep only) in `VarZeroVec`, so we can have `VarZeroVec<Vec<T>>` as long as `T` is `ULE`. It behaves like a `Vec<Vec<T>>`

Note that the `T: ULE` bound means that most of the types you use here will be of the form `VarZeroVec<Vec<PlainOldULE<N>>`, i.e. the inner vector will expose `ULE` types in the API. This is kinda annoying but not a deal breaker.

This is not a hard constraint on the design, it would be possible to improve this by creating a `OwnedULEVec<T>` type (name suggestions welcome) that is basically just a `Vec<T::ULE>` that exposes native-endian getters. This could also be used as the `Owned` variant on `ZeroVec`. Happy to make these changes if folks want; overall I feel like having folks call `.into()` when needed is mostly _fine_.

We cannot support `VarZeroVec<ZeroVec<T>>` because `ZeroVec` has a lifetime (using the `'static` variant here would not be sound).

### CODE REVIEW

File: utils/zerovec/src/ule/mod.rs [hunk 1]
```diff
@@ -8,14 +8,19 @@
...
 
 /// Fixed-width, byte-aligned data that can be cast to and from a little-endian byte slice.
 ///
 /// "ULE" stands for "Unaligned little-endian"
-pub trait ULE
+///
+/// # Safety
+///
+/// See the safety invariant documented on [`Self::from_byte_slice_unchecked()`] to implement this trait.
+pub unsafe trait ULE
```
Comment (Manishearth) [inline utils/zerovec/src/ule/mod.rs:23]: See https://twitter.com/CAD97_/status/1436399266939932691 for why this needs to be unsafe

File: utils/zerovec/src/ule/mod.rs [hunk 2]
```diff
@@ -33,8 +38,30 @@ where
...
+    /// Callers of this method must take care to ensure that `bytes` was previously passed through
+    /// [`ULE::parse_byte_slice()`] with success (and was not changed since then).
+    ///
+    /// ## Implementors
+    /// This method _must_ be implemented to return the same result as [`ULE::parse_byte_slice()`].
+    ///
+    /// Implementations of this method may involve `unsafe{}` blocks to cast the pointer to the
+    /// correct type. It is up to the implementation to reason about the safety, assuming the invariant
+    /// above.
+    unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &[Self];
```
Comment (sffc) [inline utils/zerovec/src/ule/mod.rs:63]: Why do you need to add the new trait method?  It seems orthogonal to the problem at hand, which is enabling `VZV<Vec<T>>`.

File: utils/zerovec/src/ule/vec.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::ule::*;
+
+impl<T> AsVarULE for Vec<T>
```
Comment (sffc) [inline utils/zerovec/src/ule/vec.rs:7]: Could you add a convenience impl somewhere such that if you have `VZV<Vec<T::ULE>>`, you can get a `ZV<T>` out of it?  This should be possible by getting the slice and then using the `ZV::Borrowed` variant.

**Review by sffc — CHANGES_REQUESTED** [review]

File: utils/zerovec/src/ule/mod.rs (hunk 2 above)
Comment (Manishearth) [reply utils/zerovec/src/ule/mod.rs:63]: (VarULE and ULE have two different from_byte_slice_unchecked functions)

File: utils/zerovec/src/ule/vec.rs (hunk 3 above)
Comment (Manishearth) [reply utils/zerovec/src/ule/vec.rs:7]: (Will deal with in a followup)

Comment (Manishearth) [reply utils/zerovec/src/ule/vec.rs:8]: dealing with this now, actually

**Review by sffc — APPROVED** [review]

**Comment by Manishearth** [comment]
Followup for custom derive in https://github.com/unicode-org/icu4x/issues/1044
